sw/host/opentitanlib/src/transport/cw310/usb.rs

```diff
@@ -357,6 +375,185 @@ impl Backend {
             Err(GpioError::InvalidPinName(pinname).into())
         }
     }
+
+    /// Write a byte to the CDCE906 PLL chip.
+    fn pll_write(&self, addr: u8, data: u8) -> Result<()> {
+        self.send_ctrl(Backend::CMD_PLL, 0, &[Backend::REQ_PLL_WRITE, addr, data])?;
+        let mut resp = [0u8; 2];
+        self.read_ctrl(Backend::CMD_PLL, 0, &mut resp)?;
+        if resp[0] != Backend::RESP_PLL_OK {
+            Err(
+                TransportError::PllProgramFailed(format!("CDCE906 write error: {}", resp[0]))
+                    .into(),
+            )
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Read a byte from the CDCE906 PLL chip.
+    fn pll_read(&self, addr: u8) -> Result<u8> {
+        self.send_ctrl(Backend::CMD_PLL, 0, &[Backend::REQ_PLL_READ, addr, 0])?;
+        let mut resp = [0u8; 2];
+        self.read_ctrl(Backend::CMD_PLL, 0, &mut resp)?;
+        if resp[0] != Backend::RESP_PLL_OK {
+            Err(TransportError::PllProgramFailed(format!("CDCE906 read error: {}", resp[0])).into())
+        } else {
+            Ok(resp[1])
+        }
+    }
+
+    /// Enable or disable the CDCE906 PLL chip.
+    pub fn pll_enable(&self, enable: bool) -> Result<()> {
+        let mut reg = self.pll_read(12)?;
+        if enable {
+            reg &= !(1 << 6);
+        } else {
+            reg |= 1 << 6;
+        }
+        self.pll_write(12, reg)
+    }
+
+    /// Calculate the multiply and divide values for the given frequency.
+    fn pll_calc_mul_div(&self, target_freq: u32) -> Result<PllMulDiv> {
+        const TARGET_FREQ_MIN: u32 = 630_000;
+        const TARGET_FREQ_MAX: u32 = 167_000_000;
+        if !(TARGET_FREQ_MIN..=TARGET_FREQ_MAX).contains(&target_freq) {
+            return Err(TransportError::PllProgramFailed(format!(
+                "Target frequency out of range: {}",
+                target_freq
+            ))
+            .into());
+        }
+
+        const REF_FREQ: u32 = 12_000_000;
+        const FVCO_MIN: u32 = 80_000_000;
+        const FVCO_MAX: u32 = 300_000_000;
+        let mut res = PllMulDiv {
+            ..Default::default()
+        };
+        // `outdiv` range to put `fvco` in [80 MHz, 300 MHz].
+        let outdiv_min: u8 = cmp::max(FVCO_MIN / target_freq, 1u32).try_into()?;
+        let outdiv_max: u8 = cmp::min(FVCO_MAX / target_freq, 127u32).try_into()?;
+        let mut best_err: u64 = u64::MAX;
+
+        'outer: for outdiv in outdiv_min..=outdiv_max {
+            let fvco_exp = u64::from(target_freq) * u64::from(outdiv);
```

Comment:
Nothing wrong with the `from`s, but since these conversions are infallible, you could write `foo as u64`.